server/schedulers/balance_leader.go

```diff
@@ -161,42 +164,92 @@ func (l *balanceLeaderScheduler) Schedule(cluster schedule.Cluster) []*operator.
 			targets[j].LeaderScore(leaderSchedulePolicy, jOp)
 	})
 
-	for i := 0; i < len(sources) || i < len(targets); i++ {
-		if i < len(sources) {
-			plan.source, plan.target = sources[i], nil
+	usedRegions := make(map[uint64]struct{})
+	sourcePoint := 0
+	targetPoint := 0
+	for sourcePoint < len(sources) || targetPoint < len(targets) {
+		if sourcePoint < len(sources) {
+			used := false
+			plan.source, plan.target = sources[sourcePoint], nil
 			retryLimit := l.retryQuota.GetLimit(plan.source)
 			log.Debug("store leader score", zap.String("scheduler", l.GetName()), zap.Uint64("source-store", plan.SourceStoreID()))
 			l.counter.WithLabelValues("high-score", plan.SourceMetricLabel()).Inc()
 			for j := 0; j < retryLimit; j++ {
 				schedulerCounter.WithLabelValues(l.GetName(), "total").Inc()
 				if ops := l.transferLeaderOut(plan); len(ops) > 0 {
+					if _, ok := usedRegions[ops[0].RegionID()]; ok {
+						continue
+					}
 					l.retryQuota.ResetLimit(plan.source)
 					ops[0].Counters = append(ops[0].Counters, l.counter.WithLabelValues("transfer-out", plan.SourceMetricLabel()))
-					return ops
+					result = append(result, ops...)
+					if len(result) >= l.conf.Batch {
+						return result
+					}
+					used = true
+					usedRegions[ops[0].RegionID()] = struct{}{}
+					schedule.AddOpInfluence(ops[0], plan.opInfluence, cluster)
+					sortStores(sources, sourcePoint, func(i, j int) bool {
+						iOp := plan.GetOpInfluence(sources[i].GetID())
```

Comment:
I wonder this may affect performance.  if it is not reordered, is it also work fine with a small batch?